processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/TargetReference.java

```diff
@@ -56,55 +58,70 @@
     /**
      * Builds a {@link TargetReference} from an {@code @Mappping}.
      */
-    public static class BuilderFromTargetMapping {
+    public static class Builder {
 
-        private Mapping mapping;
         private Method method;
         private FormattingMessager messager;
         private TypeFactory typeFactory;
 
+        // mapping parameters
+        private boolean isReversed = false;
+        private boolean isIgnored = false;
+        private String targetName = null;
+        private AnnotationMirror annotationMirror = null;
+        private AnnotationValue targetAnnotationValue = null;
+        private AnnotationValue sourceAnnotationValue = null;
+        private Method templateMethod = null;
         /**
          * During {@link #getTargetEntries(Type, String[])} an error can occur. However, we are invoking
          * that multiple times because the first entry can also be the name of the parameter. Therefore we keep
          * the error message until the end when we can report it.
          */
         private MappingErrorMessage errorMessage;
 
-        public BuilderFromTargetMapping messager(FormattingMessager messager) {
+        public Builder messager(FormattingMessager messager) {
             this.messager = messager;
             return this;
         }
 
-        public BuilderFromTargetMapping mapping(Mapping mapping) {
-            this.mapping = mapping;
+        public Builder mapping(Mapping mapping) {
+            if ( mapping.getInheritContext() != null ) {
+                this.isReversed = mapping.getInheritContext().isReversed();
+                this.templateMethod = mapping.getInheritContext().getInheritedFromMethod();
+            }
+            this.isIgnored = mapping.isIgnored();
+            this.targetName = mapping.getTargetName();
+            this.annotationMirror = mapping.getMirror();
+            this.targetAnnotationValue = mapping.getTargetAnnotationValue();
+            this.sourceAnnotationValue = mapping.getSourceAnnotationValue();
             return this;
         }
 
-        public BuilderFromTargetMapping method(Method method) {
-            this.method = method;
+        public Builder typeFactory(TypeFactory typeFactory) {
+            this.typeFactory = typeFactory;
             return this;
         }
 
-        public BuilderFromTargetMapping typeFactory(TypeFactory typeFactory) {
-            this.typeFactory = typeFactory;
+        public Builder method(Method method) {
+            this.method = method;
             return this;
         }
 
         public TargetReference build() {
 
-            boolean isInverse = mapping.getInheritContext() != null && mapping.getInheritContext().isReversed();
+            Objects.requireNonNull( method );
```

Comment:
Perhaps we can guarantee this by creating the builder with a constructor that takes `method`, `typeFactory` and `messager`?

Reply:
thought about that.. but abandoned that idea. It breaks the fluent api and we replace the construction of an object type problem (too many constructor args, often of the same type) to  the builder.

I don't see anything wrong with `Objects.requireNonNull` it was introduced in java7 and if you read the java doc especially for checking constructor and type args. 

The idea is to fail fast (and not somewhere inside a stream or fluent statement were you have to find which parameter actually caused the problem).

Reply:
Fine for me, the reason why I don't like it is because it throw an NPE without any message. Although indeed you would know why that happened on that line. We should rethink the internals a bit and pass one object that would return all these utils everywhere instead of passing them all the time one by one :)

Reply:
agreed..  I'll create a refactor issue for that..

Reply:
Thanks